In [1]:
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql import types

In [2]:
spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


23/03/06 17:53:51 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


### Question 1: 

**Install Spark and PySpark** 

- Install Spark
- Run PySpark
- Create a local spark session
- Execute spark.version.

What's the output?
- 3.3.2
- 2.1.4
- 1.2.3
- 5.4
</br></br>

In [3]:
pyspark.__version__

'3.3.2'

In [4]:
spark.version

'3.3.2'

### Question 2: 

**HVFHW June 2021**

Read it with Spark using the same schema as we did in the lessons.</br> 
We will use this dataset for all the remaining questions.</br>
Repartition it to 12 partitions and save it to parquet.</br>
What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.</br>

- 2MB
- 24MB
- 100MB
- 250MB

In [5]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz

--2023-03-06 17:53:54--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e-a6da-4923-ad6f-35ff02446a51?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230306%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230306T175354Z&X-Amz-Expires=300&X-Amz-Signature=31073219021e98517c88f093a9134e675f4bdf9ce2c6725c3b639fd9e739b28a&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhvhv_tripdata_2021-06.csv.gz&response-content-type=application%2Foctet-stream [following]
--2023-03-06 17:53:54--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e

In [6]:
!ls -lh ./data/raw/fhvhv/2021/06/fhvhv_tripdata_2021-06.csv.gz

-rw-rw-r-- 1 blanpa blanpa 168M Dec 20 00:13 ./data/raw/fhvhv/2021/02/fhvhv_tripdata_2021-06.csv.gz


In [7]:
# types.StructField('hvfhs_license_num', types.StringType(), True),

schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropoff_datetime', types.TimestampType(), True),
    types.StructField('PULocationID', types.IntegerType(), True),
    types.StructField('DOLocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.StringType(), True),
    types.StructField('Affiliated_base_number', types.StringType(), True),
])

In [10]:
df = spark.read \
    .option("header", "true") \
    .schema(schema) \
    .csv('./data/raw/fhvhv/2021/06/fhvhv_tripdata_2021-06.csv.gz')

df = df.repartition(12)

df.write.parquet('data/pq/fhvhv/2021/06/', mode = "overwrite") #compression="gzip"

In [11]:
df = spark.read.parquet('data/pq/fhvhv/2021/06/')

In [12]:
df.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02889|2021-06-04 20:51:44|2021-06-04 21:10:12|         239|         158|      N|                B02889|
|              B02800|2021-06-04 15:50:15|2021-06-04 16:19:29|          75|         116|      N|                  null|
|              B02510|2021-06-02 21:03:38|2021-06-02 21:10:12|         167|         168|      N|                  null|
|              B02867|2021-06-02 12:51:57|2021-06-02 13:05:09|         151|         142|      N|                B02867|
|              B02869|2021-06-21 09:51:45|2021-06-21 10:09:17|         106|          65|      N|                B02869|
|              B02764|2021-06-02 13:27:0

In [13]:
!ls -lh ./data/pq/fhvhv/2021/06/

total 284M
-rw-r--r-- 1 blanpa blanpa   0 Mar  6 18:02 _SUCCESS
-rw-r--r-- 1 blanpa blanpa 24M Mar  6 18:02 part-00000-fdf53b0e-9480-4e18-a1e1-d56c9b130648-c000.snappy.parquet
-rw-r--r-- 1 blanpa blanpa 24M Mar  6 18:02 part-00001-fdf53b0e-9480-4e18-a1e1-d56c9b130648-c000.snappy.parquet
-rw-r--r-- 1 blanpa blanpa 24M Mar  6 18:02 part-00002-fdf53b0e-9480-4e18-a1e1-d56c9b130648-c000.snappy.parquet
-rw-r--r-- 1 blanpa blanpa 24M Mar  6 18:02 part-00003-fdf53b0e-9480-4e18-a1e1-d56c9b130648-c000.snappy.parquet
-rw-r--r-- 1 blanpa blanpa 24M Mar  6 18:02 part-00004-fdf53b0e-9480-4e18-a1e1-d56c9b130648-c000.snappy.parquet
-rw-r--r-- 1 blanpa blanpa 24M Mar  6 18:02 part-00005-fdf53b0e-9480-4e18-a1e1-d56c9b130648-c000.snappy.parquet
-rw-r--r-- 1 blanpa blanpa 24M Mar  6 18:02 part-00006-fdf53b0e-9480-4e18-a1e1-d56c9b130648-c000.snappy.parquet
-rw-r--r-- 1 blanpa blanpa 24M Mar  6 18:02 part-00007-fdf53b0e-9480-4e18-a1e1-d56c9b130648-c000.snappy.parquet
-rw-r--r-- 1 blanpa blanpa 24M Mar  6 18

### Question 3: 

**Count records**  

How many taxi trips were there on June 15?</br></br>
Consider only trips that started on June 15.</br>

- 308,164
- 12,856
- 452,470
- 50,982
</br></br>

In [15]:
from pyspark.sql import functions as F

In [17]:
df \
    .withColumn('pickup_date', F.to_date(df.pickup_datetime)) \
    .filter("pickup_date = '2021-06-15'") \
    .count()

452470

In [18]:
df.registerTempTable('fhvhv_2021_06')

/home/blanpa/miniconda3/envs/dez/lib/python3.10/site-packages/pyspark/sql/dataframe.py:229: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn("Deprecated in 2.0, use createOrReplaceTempView instead.", FutureWarning)


In [20]:
spark.sql("""
SELECT
    COUNT(1)
FROM 
    fhvhv_2021_06
WHERE
    to_date(pickup_datetime) = '2021-06-15';
""").show()

+--------+
|count(1)|
+--------+
|  452470|
+--------+



### Question 4: 

**Longest trip for each day**  

Now calculate the duration for each trip.</br>
How long was the longest trip in Hours?</br>

- 66.87 Hours
- 243.44 Hours
- 7.68 Hours
- 3.32 Hours
</br></br>

In [21]:
df.columns

['dispatching_base_num',
 'pickup_datetime',
 'dropoff_datetime',
 'PULocationID',
 'DOLocationID',
 'SR_Flag',
 'Affiliated_base_number']

In [22]:
df \
    .withColumn('duration', df.dropoff_datetime.cast('long') - df.pickup_datetime.cast('long')) \
    .withColumn('pickup_date', F.to_date(df.pickup_datetime)) \
    .groupBy('pickup_date') \
        .max('duration') \
    .orderBy('max(duration)', ascending=False) \
    .limit(5) \
    .show()

+-----------+-------------+
|pickup_date|max(duration)|
+-----------+-------------+
| 2021-06-25|       240764|
| 2021-06-22|        91979|
| 2021-06-27|        71931|
| 2021-06-26|        65510|
| 2021-06-23|        59281|
+-----------+-------------+



In [29]:
spark.sql("""
SELECT
    to_date(pickup_datetime) AS pickup_date,
    MAX((CAST(dropoff_datetime AS LONG) - CAST(pickup_datetime AS LONG)) /60 /60) AS duration_in_hours
FROM 
    fhvhv_2021_06
GROUP BY
    1
ORDER BY
    2 DESC
LIMIT 10;
""").show()

+-----------+------------------+
|pickup_date| duration_in_hours|
+-----------+------------------+
| 2021-06-25| 66.87888888888888|
| 2021-06-22|25.549722222222222|
| 2021-06-27|19.980833333333333|
| 2021-06-26| 18.19722222222222|
| 2021-06-23|16.466944444444444|
| 2021-06-24|13.909722222222223|
| 2021-06-04|             11.67|
| 2021-06-20|10.984444444444446|
| 2021-06-01|           10.2675|
| 2021-06-28|  9.96638888888889|
+-----------+------------------+



**Q5**: Most frequent `dispatching_base_num`

How many stages this spark job has?

### Question 5: 

**User Interface**

 Spark’s User Interface which shows application's dashboard runs on which local port?</br>

- 80
- 443
- 4040
- 8080


In [25]:
spark.sparkContext.uiWebUrl

'http://de-zoomcamp-new.europe-west1-b.c.utility-logic-375619.internal:4040'

### Question 6: 

**Most frequent pickup location zone**

Load the zone lookup data into a temp view in Spark</br>
[Zone Data](https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv)</br>

Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?</br>

- East Chelsea
- Astoria
- Union Sq
- Crown Heights North

In [38]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
!mv taxi_zone_lookup.csv ./data/csv/misc/

--2023-03-06 18:45:41--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230306%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230306T184522Z&X-Amz-Expires=300&X-Amz-Signature=1c73036351d39f30b247777eb4f45db6a65d0842d00f9b248a4ee24801eda221&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [following]
--2023-03-06 18:45:41--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6e

In [50]:
schema = types.StructType([
    types.StructField('LocationID', types.IntegerType(), True),
    types.StructField('Borough', types.StringType(), True),
    types.StructField('Zone', types.StringType(), True),
    types.StructField('service_zone', types.StringType(), True)
])

In [52]:
df_zones = spark.read.option("header", "true").schema(schema).csv('./data/csv/misc/taxi_zone_lookup.csv')


In [53]:
df_zones.show()

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
|         6|Staten Island|Arrochar/Fort Wad...|   Boro Zone|
|         7|       Queens|             Astoria|   Boro Zone|
|         8|       Queens|        Astoria Park|   Boro Zone|
|         9|       Queens|          Auburndale|   Boro Zone|
|        10|       Queens|        Baisley Park|   Boro Zone|
|        11|     Brooklyn|          Bath Beach|   Boro Zone|
|        12|    Manhattan|        Battery Park| Yellow Zone|
|        13|    Manhattan|   Battery Park City| Yellow Zone|
|        14|     Brookly

In [54]:
df_zones.columns

['LocationID', 'Borough', 'Zone', 'service_zone']

In [55]:
df.columns

['dispatching_base_num',
 'pickup_datetime',
 'dropoff_datetime',
 'PULocationID',
 'DOLocationID',
 'SR_Flag',
 'Affiliated_base_number']

In [56]:
df_zones.registerTempTable('zones')

/home/blanpa/miniconda3/envs/dez/lib/python3.10/site-packages/pyspark/sql/dataframe.py:229: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  warnings.warn("Deprecated in 2.0, use createOrReplaceTempView instead.", FutureWarning)


In [83]:
sql = spark.sql("""
SELECT
    pul.Zone AS pu_location,
    COUNT(1) as COUNT
FROM 
    fhvhv_2021_06 fhv JOIN zones pul ON fhv.PULocationID = pul.LocationID
GROUP BY 
    1
ORDER BY
    2 DESC
LIMIT 5;
""")

In [84]:
sql.show()

+-------------------+------+
|        pu_location| COUNT|
+-------------------+------+
|Crown Heights North|231279|
|       East Village|221244|
|        JFK Airport|188867|
|     Bushwick South|187929|
|      East New York|186780|
+-------------------+------+

